index.js

```diff
@@ -85,27 +82,38 @@ module.exports = function (Adapter) {
           return error ? reject(error) : resolve(files)
         })
       })
-    ).then(function (files) {
-      return Promise.all(map(files, function (file) {
-        return new Promise(function (resolve, reject) {
-          var filePath = path.join(dbPath, type, '' + file)
+    ).then(async function (files) {
+      const allThePromises = []
+      const concurrentReads = 256
+      let promiseCount = 0
+
+      while(promiseCount < files.length) {
```

Comment:
hmm, although this works, what it's actually doing is processing batches of 256 files at once. let's say there are 1024 files to read. it will execute in order:
- 0-256
- 256-512
- 512-768
- 768-1024

might be more efficient to process a maximum of 256 files at once, [for example](https://gist.github.com/jcouyang/632709f30e12a7879a73e9e132c0d56b).

Reply:
this is a good point.  I was trying to make minimal changes to the existing code, but I have done this sort of things before, where the upper bound is defined, and you just let rip until it's full, then push a new one on each time one comes off.

I agree `256` is too high.  there is no perfect answer because the size of the records makes a big difference, but without benchmarking, I think we just pick something reasonable.  `32` is probably a good number but I will run a few benchmarks on my data

Reply:
some interesting results:

benchmark with 124732 records (files)

| concurrent read limit | execution time (ms) |
| --- | --- |
|32 | 44872|
|64 | 22324|
|128 | 11163|
|256 | 9616|
|512 | 6993|
|1024 | 6298|
|2048 | 6376|
|4096 | 6602|
|8192 | 6497|
|32768 | 6453|
|65536 | 6599|
|131072 | Error: EMFILE: too many open files|

conclusion:

- there's clearly a perf cliff beyond ~512
- this seems to suggest the sweet spot is `1024`
- surprisingly, I didn't get the open files error until specifying 131072 concurrent requests